### ISSUE TITLE
Pytest Incompatibility with ROS2 Humble Release on Robostack

### ISSUE THREAD

**Issue by alberthli** [prefix-dev/pixi#1529]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://github.com/prefix-dev/pixi/releases) of pixi, using `pixi --version`.

### Reproducible example

The issue can be reproduced with the following `pixi.toml`.
```shell
[project]
name = "proj"
channels = ["conda-forge", "robostack-staging"]
platforms = ["linux-64"]

[dependencies]
ros-humble-ros-base = "*"

[pypi-dependencies]
pytest = "*"
```

### Issue description

With the above `pixi.toml`, running `pytest` in the `pixi` shell raises the error
```
.pixi/envs/default/lib/python3.11/site-packages/pluggy/_hooks.py:513: in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
.pixi/envs/default/lib/python3.11/site-packages/pluggy/_manager.py:120: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
.pixi/envs/default/lib/python3.11/site-packages/_pytest/python.py:197: in pytest_collect_file
    module: Module = ihook.pytest_pycollect_makemodule(
.pixi/envs/default/lib/python3.11/site-packages/_pytest/config/compat.py:79: in fixed_hook
    return hook(**kw)
.pixi/envs/default/lib/python3.11/site-packages/pluggy/_hooks.py:513: in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
.pixi/envs/default/lib/python3.11/site-packages/pluggy/_manager.py:120: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
.pixi/envs/default/lib/python3.11/site-packages/launch_testing/pytest/hooks.py:188: in pytest_pycollect_makemodule
    entrypoint = find_launch_test_entrypoint(path)
.pixi/envs/default/lib/python3.11/site-packages/launch_testing/pytest/hooks.py:178: in find_launch_test_entrypoint
    module = import_path(path, root=None)
E   TypeError: import_path() missing 1 required keyword-only argument: 'consider_namespace_packages'
```
This seems to have been patched and backported to Humble here: [https://github.com/ros2/launch/pull/777](https://github.com/ros2/launch/pull/777). However, it's not clear whether the binary from RoboStack includes the patch. Running `pytest` outside of a `pixi` environment where I have ROS2 installed locally doesn't raise any errors.

### Expected behavior

I should be able to run `pytest` without issue in a `pixi` shell.

**Comment by tdejager** [comment]
This `pixi.toml` seems to work for me on `osx-arm64`. Can you see if the same issue occurs with this one? Your original is missing the `python` dependency.

```toml
[project]
name = "proj"
channels = ["conda-forge", "robostack-staging"]
platforms = ["linux-64", "osx-arm64"]

[dependencies]
python = "3.11"
ros-humble-ros-base = "*"

[pypi-dependencies]
pytest = "*"
```

Seems to be using this conda `pytest` version:

```
pytest                                             8.2.2         pyhd8ed1ab_0          251 KiB    conda  pytest-8.2.2-pyhd8ed1ab_0.conda
```

**Comment by alberthli** [comment]
My apologies. Yes, this still occurs when I add the missing `python` dependency.

```
[project]
name = "proj"
channels = ["conda-forge", "robostack-staging"]
platforms = ["linux-64"]

[dependencies]
python = ">=3.10"
ros-humble-ros-base = "*"

[pypi-dependencies]
pytest = "*"
```

**Comment by wolfv** [comment]
@alberthli - do you mean we need to update the `ros-humble-launch-testing` package for humble?

**Comment by alberthli** [comment]
I'm not 100% sure, but it might be a good start, since it seems like this exact bug was resolved about 1 month ago on the ROS2 side. If the package hasn't been updated on RoboStack since then, it's probably at least a necessary condition.

**Comment by wolfv** [comment]
Package should be online in a few minutes (max 30 minutes). Then, can you run `pixi update` and see if the issue persists?

**Comment by wolfv** [comment]
The other workaround would be to downgrade `pytest` I think.

**Comment by alberthli** [comment]
Thanks, I'll give it a shot (is there any way for me to verify on my end whether the new package has gone online yet?).

I would downgrade `pytest`, but it seems like the version is pinned on the ROS2 side. For instance, if I pin it to `8.0.0` in the `pixi.toml`, I get
```
  × failed to solve the pypi requirements of 'default' 'linux-64'
  ├─▶ failed to resolve pypi dependencies
  ╰─▶ Because you require pytest==8.0.0 and pytest==8.2.2, we can conclude that the requirements are unsatisfiable.
```
upon running `pixi shell`.

**Comment by wolfv** [comment]
In that case you probably can also remove `pytest` in your `pypi-dependencies` (if it's already installed).

**Comment by wolfv** [comment]
actually, what i did didn't work. let me see about that.

**Comment by tdejager** [comment]
> In that case you probably can also remove `pytest` in your `pypi-dependencies` (if it's already installed).

In my case it seems to get pytest from conda, the pypi-dependencies does not seem to have an effect on osx-arm64 at least, probably included by the ROS base?

**Comment by alberthli** [comment]
Interestingly, for me, whether or not I include `pytest` in my `pypi-dependencies`, `pip show pytest` always indicates that `pytest` 8.2.2 is installed in the pixi shell. I also have as the `conda` version:
```
pytest                    8.2.2              pyhd8ed1ab_0    conda-forge
```
The `conda` version is the same either way as well.

**Comment by tdejager** [comment]
> Interestingly, for me, whether or not I include `pytest` in my `pypi-dependencies`, `pip show pytest` always indicates that `pytest` 8.2.2 is installed in the pixi shell. I also have as the `conda` version:
> 
> ```
> 
> pytest                    8.2.2              pyhd8ed1ab_0    conda-forge
> 
> ```
> 
> The `conda` version is the same either way as well.

Yeah the pypi dependencies for purest don't do anything because the base includes: https://prefix.dev/channels/robostack-staging/packages/ros-humble-ament-cmake-pytest

And that pulls in the conda pytest. During the pypi resolve we use the conda version so the pypi version is never installed.

**Comment by alberthli** [comment]
OK, that makes sense. Any clue why this bug seems to affect `linux-64` but not `osx-arm64`?

My current workaround is to define some `workaround` task that downgrades `pytest` after the environment is set up, but I would love to be able to simply build the environment without issue.

**Comment by wolfv** [comment]
@alberthli I am quite sure that you can just define `pytest = 8.0` or some other older version in your dependencies (_NOT_ pypi-dependencies), and things would work (as a workaround). I am working on updating the packages in ros-humble so that this is fixed.

Try this:

```toml
[project]
name = "proj"
channels = ["conda-forge", "robostack-staging"]
platforms = ["linux-64"]

[dependencies]
python = ">=3.10"
ros-humble-ros-base = "*"
pytest = "8.0.*"  # or whatever older version works!
```

**Comment by alberthli** [comment]
Ah, thank you for the obvious solution - I completely forgot to try specifying its version as a non-pypi dependency.

**Comment by tdejager** [comment]
Do you think we could close this as this is a ros-humble issue :)?

**Comment by alberthli** [comment]
I would be OK closing it as long as I had another issue/PR that I could track - does another one exist? If not, I could open it and reference this issue in it.
